From 937eabf58e4d7e63b1867251293574aaee7358ec Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 4 May 2023 08:05:28 +0000
Subject: [PATCH 42/53] [Huawei] mm/sharepool: Extracting the mm initialization
 process from add_task

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Extracting the mm initialization process from sp_group_add_task().
There is a little change that we invoke put_task_struct() in advance.
This should make no problem since we shall not use the task_struct
after that.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 120 +++++++++++++++++++++++++++---------------------
 1 file changed, 68 insertions(+), 52 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index b8e82bb1f962..1fa872e5d5a1 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -1292,6 +1292,69 @@ static int sp_group_link_task(struct mm_struct *mm, struct sp_group *spg,
 	return 0;
 }
 
+/*
+ * Find and initialize the mm of the task specified by @tgid.
+ * We increace the usercount for the mm on success.
+ */
+static int mm_add_group_init(pid_t tgid, struct mm_struct **pmm)
+{
+	int ret;
+	struct mm_struct *mm;
+	struct task_struct *tsk;
+
+	ret = get_task(tgid, &tsk);
+	if (ret)
+		return ret;
+
+	if (check_aoscore_process(tsk)) {
+		ret = -EACCES;
+		goto out_put_task;
+	}
+
+	/*
+	 * group_leader: current thread may be exiting in a multithread process
+	 *
+	 * DESIGN IDEA
+	 * We increase mm->mm_users deliberately to ensure it's decreased in
+	 * share pool under only 2 circumstances, which will simply the overall
+	 * design as mm won't be freed unexpectedly.
+	 *
+	 * The corresponding refcount decrements are as follows:
+	 * 1. the error handling branch of THIS function.
+	 * 2. In sp_group_exit(). It's called only when process is exiting.
+	 */
+	mm = get_task_mm(tsk->group_leader);
+	if (!mm) {
+		ret = -ESRCH;
+		goto out_put_task;
+	}
+
+	if (!mm->sp_group_master) {
+		/* We don't free mm->sp_group_master on failure */
+		down_write(&sp_global_sem);
+		ret = sp_init_group_master_locked(tsk, mm);
+		up_write(&sp_global_sem);
+		if (ret)
+			goto out_put_mm;
+	}
+
+	if (mm->sp_group_master && mm->sp_group_master->tgid != tgid) {
+		pr_err("add: task(%d) is a vfork child of the original task(%d)\n",
+			tgid, mm->sp_group_master->tgid);
+		ret = -EINVAL;
+		goto out_put_mm;
+	}
+	*pmm = mm;
+
+out_put_mm:
+	if (ret)
+		mmput(mm);
+out_put_task:
+	put_task_struct(tsk);
+
+	return ret;
+}
+
 /**
  * mg_sp_group_add_task() - Add a process to an share group (sp_group).
  * @tgid: the tgid of the task to be added.
@@ -1321,7 +1384,6 @@ static int sp_group_link_task(struct mm_struct *mm, struct sp_group *spg,
 int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 {
 	struct rb_node *p;
-	struct task_struct *tsk;
 	struct mm_struct *mm;
 	struct sp_group *spg;
 	int ret = 0;
@@ -1343,55 +1405,11 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		return -EINVAL;
 	}
 
-	down_write(&sp_global_sem);
-
-	ret = get_task(tgid, &tsk);
-	if (ret) {
-		up_write(&sp_global_sem);
-		goto out;
-	}
-
-	if (check_aoscore_process(tsk)) {
-		up_write(&sp_global_sem);
-		ret = -EACCES;
-		goto out_put_task;
-	}
-
-	/*
-	 * group_leader: current thread may be exiting in a multithread process
-	 *
-	 * DESIGN IDEA
-	 * We increase mm->mm_users deliberately to ensure it's decreased in
-	 * share pool under only 2 circumstances, which will simply the overall
-	 * design as mm won't be freed unexpectedly.
-	 *
-	 * The corresponding refcount decrements are as follows:
-	 * 1. the error handling branch of THIS function.
-	 * 2. In sp_group_exit(). It's called only when process is exiting.
-	 */
-	mm = get_task_mm(tsk->group_leader);
-	if (!mm) {
-		up_write(&sp_global_sem);
-		ret = -ESRCH;
-		goto out_put_task;
-	}
-
-	if (!mm->sp_group_master) {
-		/* We don't free mm->sp_group_master on failure */
-		ret = sp_init_group_master_locked(tsk, mm);
-		if (ret) {
-			up_write(&sp_global_sem);
-			goto out_put_mm;
-		}
-	}
+	ret = mm_add_group_init(tgid, &mm);
+	if (ret < 0)
+		return ret;
 
-	if (mm->sp_group_master && mm->sp_group_master->tgid != tgid) {
-		up_write(&sp_global_sem);
-		pr_err("add: task(%d) is a vfork child of the original task(%d)\n",
-			tgid, mm->sp_group_master->tgid);
-		ret = -EINVAL;
-		goto out_put_mm;
-	}
+	down_write(&sp_global_sem);
 
 	spg = find_or_alloc_sp_group(spg_id);
 	if (IS_ERR(spg)) {
@@ -1438,9 +1456,7 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 	/* No need to put the mm if the sp group adds this mm successfully */
 	if (unlikely(ret))
 		mmput(mm);
-out_put_task:
-	put_task_struct(tsk);
-out:
+
 	return ret == 0 ? spg->id : ret;
 }
 EXPORT_SYMBOL_GPL(mg_sp_group_add_task);
-- 
2.17.1

